core/src/Cuda/Kokkos_Cuda_Instance.cpp

```diff
@@ -471,19 +457,11 @@ Cuda::size_type *CudaInternal::scratch_flags(const std::size_t size) const {
       m_scratchFlagsCount < scratch_count(size)) {
     m_scratchFlagsCount = scratch_count(size);
 
-    using Record =
-        Kokkos::Impl::SharedAllocationRecord<Kokkos::CudaSpace, void>;
-
-    if (m_scratchFlags) Record::decrement(Record::get_record(m_scratchFlags));
-
+    KOKKOS_IMPL_CUDA_SAFE_CALL(cudaSetDevice(m_cudaDev));
+    KOKKOS_IMPL_CUDA_SAFE_CALL(cudaFree(m_scratchFlags));
     std::size_t alloc_size =
         multiply_overflow_abort(m_scratchFlagsCount, sizeScratchGrain);
-    Record *const r = Record::allocate(
-        Kokkos::CudaSpace(), "Kokkos::InternalScratchFlags", alloc_size);
-
-    Record::increment(r);
-
-    m_scratchFlags = reinterpret_cast<size_type *>(r->data());
+    KOKKOS_IMPL_CUDA_SAFE_CALL(cudaMalloc(&m_scratchFlags, alloc_size));
```

Comment:
It's easier to get the device right if we don't go through `SharedAllocationRecord` which generally doesn't know about space instances and could easily set the device wrongly to the one used by the default execution space instance.